From 1c624e6430027bee09a9f05464f2e6a6bf9be54c Mon Sep 17 00:00:00 2001
From: from minimyth2 <unknown@unknown.org>
Date: Thu, 18 Jan 2024 08:23:18 +0100
Subject: [PATCH 103/184] 0377-drm-meson-swap-primary-overlay-zpos.patch

This hack fixes failed KMS atomic_commit in Qt on Amlogic. Issue seems to be related to meson-drm zpos
property reports for overlay and primarly planes.

Currnetly meson-drm driver reports zpos=1 for primarly plane - so Qt usues this zpos value at KMS
modeset atomic_commits. Unfortunatelly, with this zpos value Qt atomic_commits are failing. Error
code suggest failure is due wrong parameters provided by Qt to drmModeAtomicCommit libdrm call.

Changing primarly plane zpos value to 0 makes Qt atomic_commits working - so this may suggest mesond-drm
DRM kernel driver wrognly reposts zpos values for primarly & overlay planes.
Unfortunatelly DRM experts have different opinion about this...

Anyway, proposed hack makes Qt atomic_commits working out-of-box on all test platforms I have
(s905w, s912 and SM1).
---
 drivers/gpu/drm/meson/meson_overlay.c | 4 ++--
 drivers/gpu/drm/meson/meson_plane.c   | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/meson/meson_overlay.c b/drivers/gpu/drm/meson/meson_overlay.c
index 7f98de388..1c4bd6f35 100644
--- a/drivers/gpu/drm/meson/meson_overlay.c
+++ b/drivers/gpu/drm/meson/meson_overlay.c
@@ -855,8 +855,8 @@ int meson_overlay_create(struct meson_drm *priv)
 
 	drm_plane_helper_add(plane, &meson_overlay_helper_funcs);
 
-	/* For now, VD Overlay plane is always on the back */
-	drm_plane_create_zpos_immutable_property(plane, 0);
+	/* For now, VD Overlay plane is always on the front */
+	drm_plane_create_zpos_immutable_property(plane, 1);
 
 	priv->overlay_plane = plane;
 
diff --git a/drivers/gpu/drm/meson/meson_plane.c b/drivers/gpu/drm/meson/meson_plane.c
index 815dfe304..3d229499d 100644
--- a/drivers/gpu/drm/meson/meson_plane.c
+++ b/drivers/gpu/drm/meson/meson_plane.c
@@ -557,8 +557,8 @@ int meson_plane_create(struct meson_drm *priv)
 
 	drm_plane_helper_add(plane, &meson_plane_helper_funcs);
 
-	/* For now, OSD Primary plane is always on the front */
-	drm_plane_create_zpos_immutable_property(plane, 1);
+	/* For now, OSD Primary plane is always on the back */
+	drm_plane_create_zpos_immutable_property(plane, 0);
 
 	priv->primary_plane = plane;
 
-- 
2.39.2

